### PR TITLE
Support Applying Configuration and Get Configuration

### PR DESCRIPTION
**Pull request by sammychinedu2ky**
Fixes https://github.com/G-Research/consuldotnet/issues/268

### CODE REVIEW

File: Consul/Configuration.cs [hunk 1]
```diff
@@ -0,0 +1,1746 @@
...
+        public string MutualTLSMode { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public List<EnvoyExtensionConfig> EnvoyExtensions { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public DestinationConfig Destination { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public int MaxInboundConnections { get; set; }
```
Comment (marcin-krystianc) [inline Consul/Configuration.cs:1517]: We can make it nullable so it is not sent if it is not set.

Comment (sammychinedu2ky) [reply Consul/Configuration.cs:1517]: Alright

File: Consul.Test/ConfigurationTest.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Consul.Test
+{
+    public class ConfigurationTest : BaseFixture
+    {
+        [Fact]
+        public async Task Configuration_ApplyConfig()
```
Comment (marcin-krystianc) [inline Consul.Test/ConfigurationTest.cs:32]: We are modifying the agent's configuration which is a global state. So Is it okay, to run this test simultaneously with other tests? If it is not ok then we need to mark this test with `ExclusiveCollection` attribute.

**Review by marcin-krystianc — COMMENTED** [review]
I think it would make a lot of sense if you implement https://developer.hashicorp.com/consul/api-docs/config#get-configuration, so we can use it in the test to assert that the configuration has been applied.

File: Consul.Test/ConfigurationTest.cs (hunk 2 above)
Comment (sammychinedu2ky) [reply Consul.Test/ConfigurationTest.cs:32]: okay

**Comment by sammychinedu2ky** [comment]
> I think it would make a lot of sense if you implement https://developer.hashicorp.com/consul/api-docs/config#get-configuration, so we can use it in the test to assert that the configuration has been applied.

@marcin-krystianc I should also implement get-configuration in this PR?

**Comment by sammychinedu2ky** [comment]
Done @marcin-krystianc

**Comment by sammychinedu2ky** [comment]
Hi @marcin-krystianc. It has been resolved 🚀

File: Consul.Test/ConfigurationTest.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+using Xunit;
+
+namespace Consul.Test
+{
+    [CollectionDefinition("NonParallelCollection")]
+    public class NonParallelCollection : ICollectionFixture<ConfigurationTest>
+    {
+    }
+
+    [Collection("NonParallelCollection")]
```
Comment (marcin-krystianc) [inline Consul.Test/ConfigurationTest.cs:34]: I don't see a reason for defining a new collection, you can use the existing collection `[Collection(nameof(ExclusiveCollection))]`.

File: Consul.Test/test_config.json [hunk 4]
```diff
@@ -1,15 +1,16 @@
...
+  "acl": {
+    "enabled": true,
+    "default_policy": "allow",
+    "tokens": {
+      "master": "c6402d2b-f081-45ef-a0f2-73a4a812e9b1"
+    }
+  },
+  "enable_script_checks": true,
+  "encrypt": "d8wu8CSUrqgtjVsvcBPmhQ==",
+  "enable_central_service_config":  true
```
Comment (marcin-krystianc) [inline Consul.Test/test_config.json:15]: Is this change required for the `ApplyConfig` to work?

File: Consul.Test/ConfigurationTest.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    public class ConfigurationTest : BaseFixture
+    {
+        [Fact]
+        public async Task Configuration_ApplyConfig()
+        {
+            var payload = new ServiceDefaultsEntry
+            {
+                Kind = "service-defaults",
+                Name = "web",
+                Protocol = "http"
```
Comment (marcin-krystianc) [inline Consul.Test/ConfigurationTest.cs:44]: I would parameterize the "protocol" parameter in this test, so the test runs twice, each time with a different name. That would give us more confidence that a new configuration is applied each time.

File: Consul/Interfaces/IConfigurationEndpoint.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Consul.Interfaces
+{
+    public interface IConfigurationEndpoint
+    {
+        Task<WriteResult> ApplyConfig<TConfig>(int? cas, TConfig configurationEntry, CancellationToken ct = default) where TConfig : IConfigurationEntry;
+        Task<WriteResult> ApplyConfig<TConfig>(string dc, int? cas, string ns, TConfig configurationEntry, WriteOptions q, CancellationToken ct = default) where TConfig : IConfigurationEntry;
```
Comment (marcin-krystianc) [inline Consul/Interfaces/IConfigurationEndpoint.cs:30]: Instead of explicit `dc` and 'ns` parameters we use ` WriteOptions q` (or `QueryOptions q` for reading), see other API methods.

File: Consul.Test/ConfigurationTest.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            };
+            var writeResult = await _client.Configuration.ApplyConfig(payload);
+            Assert.Equal(HttpStatusCode.OK, writeResult.StatusCode);
+            var queryResult = await _client.Configuration.GetConfig<ServiceDefaultsEntry>(payload.Kind, payload.Name);
+            Assert.Equal(HttpStatusCode.OK, queryResult.StatusCode);
+            Assert.Equal(payload.Name, queryResult.Response.Name);
+            Assert.Equal(payload.Name, queryResult.Response.Name);
```
[the remark this answers is not in the document]
Comment (sammychinedu2ky) [reply Consul.Test/ConfigurationTest.cs:51]: ooh

File: Consul.Test/ConfigurationTest.cs (hunk 3 above)
Comment (sammychinedu2ky) [reply Consul.Test/ConfigurationTest.cs:34]: alright

File: Consul.Test/test_config.json (hunk 4 above)
Comment (sammychinedu2ky) [reply Consul.Test/test_config.json:15]: yes according to the docs

File: Consul/Interfaces/IConfigurationEndpoint.cs (hunk 6 above)
Comment (sammychinedu2ky) [reply Consul/Interfaces/IConfigurationEndpoint.cs:30]: okay

**Comment by sammychinedu2ky** [comment]
resolved @marcin-krystianc

File: Consul/Configuration.cs [hunk 8]
```diff
@@ -0,0 +1,1736 @@
...
+        }
+
+        /// <summary>
+        ///  This creates or updates the given config entry.
+        /// </summary>
+        /// <param name="configurationEntry">The configuration entry</param>
+        /// <returns>An empty write result</returns>
+        public Task<WriteResult> ApplyConfig<TConfig>(TConfig configurationEntry, CancellationToken ct = default) where TConfig : IConfigurationEntry
+        {
+            var req = _client.Put("/v1/config", configurationEntry, WriteOptions.Default);
```
Comment (marcin-krystianc) [inline Consul/Configuration.cs:1694]: You can call the other variant of this method from here, so we avoid repeating the code.

File: Consul/Configuration.cs [hunk 9]
```diff
@@ -0,0 +1,1736 @@
...
+        /// This Retrieves the given config entry.
+        /// </summary>
+        /// <typeparam name="TConfig"></typeparam>
+        /// <param name="kind">The kind of config entry</param>
+        /// <param name="name">The name of config entry</param>
+        /// <param name="ct">Cancellation Token</param>
+        /// <returns>A config entry</returns>
+        public Task<QueryResult<TConfig>> GetConfig<TConfig>(string kind, string name, CancellationToken ct = default) where TConfig : IConfigurationEntry
+        {
+            var req = _client.Get<TConfig>($"/v1/config/{kind}/{name}", QueryOptions.Default);
```
Comment (marcin-krystianc) [inline Consul/Configuration.cs:1723]: You can call the other variant of this method from here, so we avoid repeating the code.

**Review by marcin-krystianc — COMMENTED** [review]
Looks good, just some minor comments about code duplication.

**Comment by sammychinedu2ky** [comment]
Hi @marcin-krystianc . for some reason I don't know why the agentmetric code fails

**Comment by sammychinedu2ky** [comment]
For some reason no issues in this build 😅 @marcin-krystianc

**Comment by marcin-krystianc** [comment]
> For some reason no issues in this run 😅 @marcin-krystianc

yeah, there is some flakiness in the agen-metric tests. We will observe it and if it happens too often it will need to be fixed.

**Review by marcin-krystianc — APPROVED** [review]
